### PR TITLE
Bug#8489 api root path

### PR DESCRIPTION
**Pull request by dmargery**
Add support for a X-Api-Root-Path header to support getting requests for the API at urls such as https://proxy.fed4fire.eu/proxies/grid5000

David

### CODE REVIEW

File: lib/grid5000/router.rb [hunk 1]
```diff
@@ -47,29 +54,27 @@ def uri_to(request, path, in_or_out = :in, relative_or_absolute = :relative)
...
+        uri = File.join("/", *[root_path, api_version, path_prefix, mounted_path].compact)
         uri = "/" if uri.blank?
         # abasu / dmargery - bug ref 7360 - for correct URI construction
         if in_or_out == :out || relative_or_absolute == :absolute
-	  root_uri=URI(base_uri(in_or_out))
-	  if root_uri.path.blank?
-	    root_path=''
+	        root_uri=URI(base_uri(in_or_out))
+	        if root_uri.path.blank?
+	          root_path=''
```
Comment (f-didier) [inline lib/grid5000/router.rb]: You seemed to have introduced a tab or too much extra spaces

File: spec/lib/grid5000/router_spec.rb [hunk 2]
```diff
@@ -41,6 +62,43 @@
...
 
+  it "Should properly combine all X-API headers supported" do
+    request = mock(Rack::MockRequest, :env => {      
+      'HTTP_X_API_ROOT_PATH' => 'proxies/grid5000',
+      'HTTP_X_API_VERSION' => 'sid',
+      'HTTP_X_API_PATH_PREFIX' => 'g5k-api'
+    })
+    Grid5000::Router.uri_to(request, "/sites/rennes/jobs").should == "/proxies/grid5000/sid/g5k-api/sites/rennes/jobs"
+  end
+
```
Comment (f-didier) [inline spec/lib/grid5000/router_spec.rb]: Mount-path is missing if we want to use all headers, as the title of the test above suggests

File: spec/lib/grid5000/router_spec.rb [hunk 3]
```diff
@@ -41,6 +62,43 @@
...
+  it "Should properly combine X-API-[Mount-Path,Version,Path-Prefix] headers" do
+    request = mock(Rack::MockRequest, :env => {      
+      'HTTP_X_API_ROOT_PATH' => 'sites/fr/grid5000',
+      'HTTP_X_API_MOUNT_PATH' => '/sites/',
+      'HTTP_X_API_VERSION' => 'sid',
+      'HTTP_X_API_PATH_PREFIX' => 'g5k-api'
+    })
+    Grid5000::Router.uri_to(request, "/sites/rennes/jobs").should == "/sites/fr/grid5000/sid/g5k-api/rennes/jobs"
+  end
+
```
Comment (f-didier) [inline spec/lib/grid5000/router_spec.rb]: The title of the test should mention Root-path header too. This test might be equivalent to the first test (see above)

File: lib/grid5000/router.rb [hunk 4]
```diff
@@ -14,9 +14,10 @@
 
 module Grid5000
   # Computes the URI to a specific path.
-  # Takes into account the X-Api-Path-Prefix (additional prefix to add to URI
-  # path), and X-Api-Mount-Path (subset of the API path to take out of the URI
-  # path).
+  # X-Api-root-Path is the entry point for all versions of the API
+  # X-Api-version is the version string by which the server is reached
+  # X-Api-Path-Prefix is the entry point of this server for a given version
```
Comment (f-didier) [inline lib/grid5000/router.rb]: I would have written "the entry point on this server" but I'm not sure

**Comment by dmargery** [comment]
There seem indeed conflicting tests.

The RootController test says that in the presence of API_VERSION but not of MOUNT_PATH, the link to the root of the service should be /sid/

But the SiteController test says that /sites, in presence of MOUNT_PATH as /sites, and API_VERSION, should present the root as /sid

This is not very clean a probably implies some rewriting at frontend level. I'll look at whether we're applying any magic inside Grid'5000 before suggesting a additional patch

David

**Comment by dmargery** [comment]
@f-didier Finally chose to keep the trailing / in all cases,

File: lib/grid5000/router.rb [hunk 5]
```diff
@@ -16,7 +16,11 @@ module Grid5000
   # Computes the URI to a specific path.
   # X-Api-root-Path is the entry point for all versions of the API
   # X-Api-version is the version string by which the server is reached
-  # X-Api-Path-Prefix is the entry point of this server for a given version
+  # X-Api-Path-Prefix is the entry point to access this service for a
+  #   given version. To use if a server does not wish to show the top level
+  #   hieararchy exposed by the services, for example to only publish the
+  #   resources of a single site (https://rennes.g5k/ ony giving access to
```
Comment (f-didier) [inline lib/grid5000/router.rb]: Typo on "ony"

File: spec/controllers/root_controller_spec.rb [hunk 6]
```diff
@@ -39,8 +39,8 @@
     }
   end
   
-  it "should correcly add the api-prefix if any" do
-    @request.env['HTTP_X_API_PATH_PREFIX'] = 'sid'
+  it "should correcly add the version if any" do
+    @request.env['HTTP_X_API_VERSION'] = 'sid'
```
Comment (f-didier) [inline spec/controllers/root_controller_spec.rb]: The commit message may explain why you replaced HTTP_X_API_PATH_PREFIX by HTTP_X_API_VERSION in this file. Was it an old mistake that you corrected here ?

**Comment by dmargery** [comment]
Missed a commit in my rebase : do not merge

**Comment by dmargery** [comment]
The merge request has now been rebased cleanly. @f-didier, please review this final version.

**Comment by f-didier** [comment]
Some tests seem to be present two times in file router_spec.rb.

**Comment by dmargery** [comment]
Exact : removed them

**Comment by f-didier** [comment]
The code is ok for me

**Comment by dmargery** [comment]
Thank you for your review

**Pull request merged**